NuRadioReco/utilities/trace_utilities.py

```diff
@@ -111,3 +112,53 @@ def get_electric_field_energy_fluence(electric_field_trace, times, signal_window
         f_signal -= f_noise * np.sum(signal_window_mask) / np.sum(noise_window_mask)
 
     return f_signal * dt * conversion_factor_integrated_signal
+
+def upsampling_fir(trace, original_sampling_frequency, int_factor=2, ntaps=2**7):
+    """
+    This function performs an upsampling by inserting a number of zeroes
+    between samples and then applying a finite impulse response (FIR) filter.
+
+    Parameters
+    ----------
+    trace: array of floats
+        Trace to be upsampled
+    original_sampling_frequency: float
+        Sampling frequency of the input trace
+    int_factor: integer
+        Upsampling factor. The resulting trace will have a sampling frequency
+        int_factor times higher than the original one
+    ntaps: integer
+        Number of taps (order) of the FIR filter
+
+    Returns
+    -------
+    upsampled_trace: array of floats
+        The upsampled trace
+    """
+
+    if (np.abs(int(int_factor) - int_factor) > 1e-3):
+        warning_msg  = "The input upsampling factor does not seem to be close to an integer."
+        warning_msg += "It has been rounded to {}".format(int(int_factor))
+        logger.warning(warning_msg)
+
+    int_factor = int(int_factor)
+
+    if (int_factor <= 1):
+        error_msg = "Upsampling factor is less or equal to 1. Upsampling will not be performed."
+        raise ValueError(error_msg)
+
+    n_zeroes = int_factor - 1
+    zeroed_trace = []
+    for point in trace[:-1]:
+        zeroed_trace.append(point)
+        for zero in range(n_zeroes):
+            zeroed_trace.append(0.)
+
+    upsampled_delta_time = 1 / (int_factor * original_sampling_frequency)
+    upsampled_times = np.arange(0, len(zeroed_trace) * upsampled_delta_time, upsampled_delta_time)
+
+    cutoff = 1./int_factor
+    fir_coeffs = firwin(ntaps, cutoff)
```

Comment:
Does the window choice matter here? Probably not very much but I think the way we've traditionally upsampled (zero-pad in frequency domain) is more akin to a rectangular window if I understand how firwin works.

Reply:
I have no experience working with FIRs, so I've just used the default values. I can change it to `window=boxcar` if you like.

Reply:
But yeah, I think that if we want the actual taps we need to use 'boxcar'. I didn't even realise. Let me try.

Reply:
Well, waveforms look ok with the boxcar window. Might as well keep it.
![Figure_1](https://user-images.githubusercontent.com/45565159/80232236-dc52a500-8654-11ea-8635-e8e60eb75233.png)

Reply:
yeah, I I'm not sure what the best choice is!  I guess if the window is big enough it might not matter.